File: go/DataStructures/Queue/Queue.go
package Queue

import "errors"

type Node struct {
	data int
	next *Node
}

func NewNode(value int) *Node {
    return &Node{
        data: value,
        next: nil,
    }
}

type Queue struct {
    head *Node
    tail *Node
}

func NewQueue() *Queue {
    return &Queue{
        head: nil,
        tail: nil,
    }
}

func (q *Queue) Enqueue(value int) {
    newNode := NewNode(value)

    if q.tail == nil {
        q.head = newNode
        q.tail = newNode
    } else {
        q.tail.next = newNode
        q.tail = newNode
    }
}

func (q *Queue) Dequeue() error {
    if q.IsEmpty() {
        return errors.New("Empty queue. Can not perform Dequeue")
    }
    if q.head.next == nil {
        q.head = nil
        q.tail = nil
    } else {
        q.head = q.head.next
    }

    return nil
}

func (q *Queue) Peek() int {
    return q.head.data
}

func (q *Queue) IsEmpty() bool {
    return q.head == nil
}

File: javascript/DataStructures/Queue.js
class Node {
    constructor(value) {
        this.data = value;
        this.next = null;
    }
}

class Queue {
    constructor() {
        this.head = null;
        this.tail = null;
    }

    enqueue(value) {
        let newNode = new Node(value);

        if (this.head === null) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
    }

    dequeue() {
        if (this.isEmpty()) {
            throw new Error("Empty queue. Cannot perform pop!")
        }
        if (this.head.next === null) {
            this.head = null;
            this.tail = null;
        } else {
            this.head = this.head.next;
        }
    }

    isEmpty() {
        return this.head === null;
    }

    peek() {
        return this.head.data;
    }
}
